assignment 6A/IntSet.java

```diff
@@ -0,0 +1,140 @@
+		for (int firstIndex = 0; firstIndex < set.integerArray.length; firstIndex++) {
+			isSet = false;
+			for (int secondIndex = 0; secondIndex < this.integerArray.length; secondIndex++) {
+				if (set.integerArray[firstIndex] == this.integerArray[secondIndex]) {
+					isSet = true;
+				}
+			}
+			if (!isSet) { return false; }
+		}
+		return true;
+	}
+
+	/**
+	 * This method find the complement of set assumming 1 to 1000 as universal set
+	 * @return complement set
+	 */
+	public IntSet getComplement() {
+
+		List<Integer> complementSetList = new ArrayList<Integer>();
+		int complementSetIndex = 0;
+		boolean elementPresent;
+		for (int firstIndex = 0; firstIndex < MAX_SET_SIZE; firstIndex++) {
+			elementPresent = false;
+			for (int secondIndex = 0; secondIndex < this.integerArray.length; secondIndex++) {
+				if ((firstIndex + 1) == this.integerArray[secondIndex]) {
+					elementPresent = true;
+					break;
+				}
+			}
+			if (!elementPresent) {
+				complementSetList.add(firstIndex+1);
+				complementSetIndex++;
+			}
+		}
+		int[] complementSet = new int[complementSetList.size()];
+		for(int listIndex =0; listIndex < complementSetList.size(); listIndex++){
+			complementSet[listIndex] = complementSetList.get(listIndex);
+		}
+		return new IntSet(complementSet);
+	}
+
+	/**
+	 * This method find the union of to sets
+	 * @param firstSet  integer set containing integer between 1 to 1000
+	 * @param secondSet integer set containing integer between 1 to 1000
+	 * @return union of two sets
+	 */
+	public static IntSet union(IntSet firstSet, IntSet secondSet) {
+		int totalElementInSet = 0;
+		for (int firstIndex = 0; firstIndex < firstSet.integerArray.length; firstIndex++) {
+			for (int secondIndex = 0; secondIndex < secondSet.integerArray.length; secondIndex++) {
+				if (firstSet.integerArray[firstIndex] == secondSet.integerArray[secondIndex]) {
+					totalElementInSet++;
+				}
+			}
+		}
+		totalElementInSet = (firstSet.integerArray.length + secondSet.integerArray.length) - totalElementInSet;
+
+		int[] unionOfSet = new int[totalElementInSet];
+		int unionSetIndex = 0;
+		for (int firstIndex = 0; firstIndex < firstSet.integerArray.length; firstIndex++) {
+			unionOfSet[unionSetIndex] = firstSet.integerArray[firstIndex];
+			unionSetIndex++;
+		}
+		for (int firstIndex = 0; firstIndex < secondSet.integerArray.length; firstIndex++) {
+			boolean elementPresent = false;
+			
+			for (int secondIndex = 0; secondIndex < unionSetIndex; secondIndex++) {
+				if (unionOfSet[secondIndex] == secondSet.integerArray[firstIndex]) { elementPresent = true; }
+			}
+			
+			if (!elementPresent) {
+				unionOfSet[unionSetIndex] = secondSet.integerArray[firstIndex];
+				unionSetIndex++;
+			}
+		}
+		return new IntSet(unionOfSet);
+	}
+
+	/**
+	 * This method return the set as single dimension array
+	 * @return single dimension integer array containing the set
+	 */
+	public int[] getSetAsArray() {
+		IntSet set = new IntSet(this.integerArray);

```

Comment:
What is the purpose of this method? if used to get array then this approach is wrong. Here you are creating the another instance of data type but you should return the current set data.

Reply:
the purpose of this method is to return the set as a array not as the abstract immutable data type IntSet.
the reason i created new object because if we don't create new object than the array of the object will be link by reference to the returning array and when we change any index in that array than original array will be changed.
Attaching demo program- -

public class Demo {

	public static void main(String[] args) {
		// demo set
		int[] array1 = new int[] {1,2,3,4,5,6,7,8,9,10};
		//created a set object
		IntSet set1 = new IntSet(array1);
		//getting a new array using method which create new object first
		int[] array2 = set1.getSetAsArray();
		//printing the array
		for(int i =0; i< array2.length; i++){
			System.out.print(array2[i]+" ");
		}
		/**
		 * output of the array will be 1 2 3 4 5 6 7 8 9 10 
		 */
		//getting the same array using a new method which does not create a new array
		int[] array3 = set1.getSetArray();
		//changing value at index 0 in array
		array3[0] = 3;
		System.out.println();
		//printing the old array again
		for(int i=0; i<array2.length; i++){
			System.out.print(array2[i]+" ");
		}
		/**
		 * output of the old array will be 3 2 3 4 5 6 7 8 9 10 
		 * 
		 * so here as we can see the value of the old array is changed because we do not create the new object
		 */
	}

}

Reply:
If there is array reference issue then you should use another approach like copy array or something else. By this way you are creating unnecessary object which allocate complete memory for that instance. So that is unnecessary also in this approach you are getting array of another object not itself.

Reply:
ok sir